# Подготовка анализа для сервиса читателей

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Введение" data-toc-modified-id="Введение-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Введение</a></span></li><li><span><a href="#Цели-работы" data-toc-modified-id="Цели-работы-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Цели работы</a></span></li><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Описание данных</a></span></li><li><span><a href="#Библиотеки" data-toc-modified-id="Библиотеки-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Библиотеки</a></span></li><li><span><a href="#Соединение" data-toc-modified-id="Соединение-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Соединение</a></span><ul class="toc-item"><li><span><a href="#Коннектор." data-toc-modified-id="Коннектор.-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Коннектор.</a></span></li><li><span><a href="#Функции-выгрузки" data-toc-modified-id="Функции-выгрузки-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Функции выгрузки</a></span></li></ul></li><li><span><a href="#Задачи" data-toc-modified-id="Задачи-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Задачи</a></span><ul class="toc-item"><li><span><a href="#Задача-1:-Сколько-книг-вышло-после-1-января-2000-года?" data-toc-modified-id="Задача-1:-Сколько-книг-вышло-после-1-января-2000-года?-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>Задача-1: Сколько книг вышло после 1 января 2000 года?</a></span></li><li><span><a href="#Задача-2:-Количество-обзоров-для-каждой-книги--и-среднюю-оценку" data-toc-modified-id="Задача-2:-Количество-обзоров-для-каждой-книги--и-среднюю-оценку-6.2"><span class="toc-item-num">6.2&nbsp;&nbsp;</span>Задача-2: Количество обзоров для каждой книги  и среднюю оценку</a></span></li><li><span><a href="#Задача-3:-Определим-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-мы-исключим-из-анализа-брошюры;" data-toc-modified-id="Задача-3:-Определим-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-мы-исключим-из-анализа-брошюры;-6.3"><span class="toc-item-num">6.3&nbsp;&nbsp;</span>Задача-3: Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры;</a></span></li><li><span><a href="#Задача-4:-Автор-с-самой-высокой-средней-оценкой-книг-—-только-книги-с-50-и-более-оценками" data-toc-modified-id="Задача-4:-Автор-с-самой-высокой-средней-оценкой-книг-—-только-книги-с-50-и-более-оценками-6.4"><span class="toc-item-num">6.4&nbsp;&nbsp;</span>Задача-4: Автор с самой высокой средней оценкой книг — только книги с 50 и более оценками</a></span></li><li><span><a href="#Задача-5:-Среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок" data-toc-modified-id="Задача-5:-Среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок-6.5"><span class="toc-item-num">6.5&nbsp;&nbsp;</span>Задача-5: Среднее количество обзоров от пользователей, которые поставили больше 50 оценок</a></span></li></ul></li></ul></div>

## Введение
Принято решение приобрести стартап - сервис для читателей книг по подписке. Мы хотим проанализировать базу данных дабы понять какие имеются специфические паттерны в мире литературных изысков. С базой данных с информацией о книгах, авторах, издательствах и рейтингах мы можем отследить какие могут быть преимущества, которыми нам можно воспользоваться.

## Цели работы
В данной работе мы нацелены достичь следующее:
1. Разграничить горизонты нами исследуемых книг - изучим количественные показатели книг вдоль временной прямой:
    <br>a. Определим эту временную границу;
    <br>b. Нужно ли нам проанализировать все данные, или же достаточно ограничиться теми, что после этого горизонта;
2. Определить какие могут быть отдельные популярные и непопулярные жанры, авторы и сами книги, чтобы понять, каким образом должна строиться стратегия достижения ценностного предложения:
    <br>a. Определим какие средние оценки у каждой книги, чтобы ответить на вопрос о самой популярной книге;
3. Отобразить возможности издательств:
    <br>a. Определим границей, при которой ту или иную книгу можем принимать всерьез, то есть исключим объекты с 50 страницами и меньше;
    <br>b. Построим график, где будут отображаться издавтельства и количество книг, изданных в них;
4. Определить автора с самой высокой оценкой книг, при этом учесть размер того, что мы можем принимать за книгу из предыдущего шага;
5. Определим среднее количество обзоров пользователей:
    <br>a. Определим границей, при которой того или иного читателя можем принимать всерьез объект с 50 страницами;
    <br>b. Построим график, где будут отображаться количество обзоров;

## Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

## Библиотеки

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
import seaborn as sns

## Соединение
### Коннектор.

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
 'pwd': 'Sdf4$2;d-d30pp', # пароль
 'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
 'port': 6432, # порт подключения
 'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

### Функции выгрузки

In [3]:
def query_window(q):
    df = (pd.io.sql.read_sql(q, con = engine))
    return df

## Задачи
### Задача-1: Сколько книг вышло после 1 января 2000 года?

In [4]:
query_window('''

SELECT 
COUNT(DISTINCT(book_id))
FROM books
WHERE 
publication_date>'2000-01-01'





''')

count
0    819

Ответ: В нашей базе данных книг имеется **819** уникальных объектов, изданных после 1 января 2000 года.

### Задача-2: Количество обзоров для каждой книги  и среднюю оценку

In [116]:
query_window('''
SELECT 
    sub_1.title_name,
    sub_2.avg_rat AS average_rating,
    sub_3.rev_num AS number_of_reviews
FROM 
    (
    SELECT
    title AS title_name,
    book_id as b_id_s1
    FROM books
    ) AS sub_1
    
    LEFT JOIN 
    
    (
    SELECT
        DISTINCT(book_id) AS b_id_s2,
        AVG(rating) AS avg_rat
    FROM
        ratings
    GROUP BY b_id_s2
    )
    AS sub_2 ON sub_1.b_id_s1 = sub_2.b_id_s2
    LEFT JOIN
    (
    SELECT 
    DISTINCT(book_id) AS b_id_s3,
    COUNT(DISTINCT(review_id)) AS rev_num
    FROM reviews
    GROUP BY b_id_s3
    
    ) AS sub_3 ON sub_1.b_id_s1 = sub_3.b_id_s3
    ORDER BY number_of_reviews DESC
    LIMIT 10
    
''')

title_name  average_rating  \
0  Disney's Beauty and the Beast (A Little Golden...        4.000000   
1                          Essential Tales and Poems        4.000000   
2                               Leonardo's Notebooks        4.000000   
3                            The Natural Way to Draw        3.000000   
4   The Cat in the Hat and Other Dr. Seuss Favorites        5.000000   
5    Anne Rice's The Vampire Lestat: A Graphic Novel        3.666667   
6                            Twilight (Twilight  #1)        3.662500   
7                          Outlander (Outlander  #1)        4.125000   
8                                      The Alchemist        3.789474   
9  Harry Potter and the Chamber of Secrets (Harry...        4.287500   

   number_of_reviews  
0                NaN  
1                NaN  
2                NaN  
3                NaN  
4                NaN  
5                NaN  
6                7.0  
7                6.0  
8                6.0  
9                6.0

Да, чем больше отзывов, тем более вероятно, что у книги будет высокая оценка.

### Задача-3: Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры;

In [67]:
query_window('''

SELECT
    DISTINCT(publishers.publisher) AS pub_name,
    max
FROM
    (SELECT 
        p_id,
        MAX(sum)
    FROM (
            SELECT 
                DISTINCT(publisher_id) AS p_id,
                SUM(num_1) as sum
            FROM (
                    SELECT 
                    DISTINCT(publisher_id),
                    COUNT(DISTINCT(book_id)) AS num_1
                    FROM books
                    WHERE num_pages>50
                    GROUP BY publisher_id) AS sub_0


            GROUP BY num_1, publisher_id) AS sub_1
    GROUP BY p_id
    ORDER BY max DESC
    LIMIT 1 ) AS sub_2 LEFT JOIN publishers ON publishers.publisher_id = sub_2.p_id
GROUP BY pub_name, p_id, max

''')

pub_name   max
0  Penguin Books  42.0

### Задача-4: Автор с самой высокой средней оценкой книг — только книги с 50 и более оценками

In [146]:
query_window('''
SELECT
    authors.author,
    sub_2.average_rating,
    sub_1.count
FROM 
    (
    SELECT
        sub_0.book_id,
        books.author_id,
        sub_0.count
    FROM
        (
        SELECT
            DISTINCT(book_id),
            COUNT(DISTINCT(rating_id))
        FROM ratings
        GROUP BY book_id
        ) AS sub_0 
        LEFT JOIN books ON books.book_id=sub_0.book_id
    WHERE count>50
    
    ) AS sub_1 LEFT JOIN authors ON authors.author_id = sub_1.author_id
    
    LEFT JOIN 
    
    (
    SELECT
    DISTINCT(book_id),
    AVG(rating) AS average_rating
    FROM ratings
    GROUP BY book_id
    ) AS sub_2 ON sub_2.book_id = sub_1.book_id
ORDER BY average_rating DESC
LIMIT 1


''')

author  average_rating  count
0  J.K. Rowling/Mary GrandPré        4.414634     82

Дж.К.Роулинг, Гарри Поттер - лучшая книга всех времен (так судят наши критики и пользователи).

### Задача-5: Среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [100]:
query_window('''
SELECT
    sub_2.user_name,
    sub_2.num_rating,
    sub_3.review_numbers
FROM 
    (
    SELECT
        user_name,
        num_rating
    FROM 
        (
        SELECT
            user_name,
            num_rating
        FROM
                (
                SELECT 
                    DISTINCT(username) AS user_name,
                    COUNT(DISTINCT(rating_id)) AS num_rating
                FROM ratings
                GROUP BY username
                )
                AS sub_0
        WHERE sub_0.num_rating>50
        ) AS sub_1
        ) AS sub_2
    
    LEFT JOIN 
    
        (
    SELECT
        username,
        AVG(rev_num) AS review_numbers
    FROM 
        (
                SELECT
                username,
                COUNT(DISTINCT(review_id)) AS rev_num
                FROM reviews
                GROUP BY username
                ) AS sub_00
            GROUP BY username
        ) AS sub_3 ON sub_3.username = sub_2.user_name
''')

user_name  num_rating  review_numbers
0  jennifermiller          53            25.0
1      martinadam          56            27.0
2          paul88          56            22.0
3       richard89          55            26.0
4     sfitzgerald          55            28.0
5          xdavis          51            18.0

А вот и наши самые активные пользователи, со своими показателями активности.